"Annotate" exported object to fix named / namespace imports of our API in Node ESM

jakebailey opened the pull request:
Our current API bundle looks like this:

```js
var ts = (() => { ... })();
if (typeof module !== "undefined" && module.exports) {
	module.exports = ts;
}
```

[cjs-module-lexer](https://github.com/nodejs/cjs-module-lexer) cannot understand this. As such, this code fails:

```js
import { createSourceFile } from "typescript";
```

```
file:///home/jabaile/work/TypeScript/complain.mjs:2
    createSourceFile,
    ^^^^^^^^^^^^^^^^
SyntaxError: Named export 'createSourceFile' not found. The requested module 'typescript' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'typescript';
```

This has been the case with TypeScript for the entire time that Node's ESM support has been around. Even before 5.0, it didn't work since our exported object was still complicated.

However, there's a footgun; this code _will not_ error:

```js
import * as ts from "typescript";
```

In TypeScript code, we'll emit helpers that fix this up. But in native ESM, this will only be an object with a `default` property! You'll write `ts.createSourceFile`, and it'll be undefined and fail later.

To fix this, we need to make `cjs-module-lexer` aware of our exports. When emitting `cjs`, `esbuild` uses a trick to "declare" the names to Node:

```js
module.exports = someComplicatedThing();
0 && (module.exports = {
	foo,
	bar,
});
```

This is dead code, but Node will still read the names.

~If we do this same thing at the bottom of our custom bundle, we can also get named imports working. Unfortunately, there's no static way within esbuild for us to know the names via the API (likely, https://github.com/evanw/esbuild/issues/3110 or https://github.com/evanw/esbuild/issues/3281; I also filed https://github.com/evanw/esbuild/issues/3607), so we have to import the bundle and grab the names. But, once we have them, we can tack the output onto the bundle.~

To fix this, the PR now instead emits CJS, then uses the header/footer (instead of just a footer + IIFE) to do something like UMD. This lets us use esbuild to make the annotations.

This now means that _both_ of these imports work!

```js
import * as ts from "typescript";
import { createSourceFile } from "typescript";

ts.createSourceFile(...);
```

Fixes #56366

Related: #54018

jakebailey commented:
@typescript-bot pack this

jakebailey commented:
@typescript-bot perf test this

fatcerberus commented:
> But in native ESM, this is the empty object! You'll write ts.createSourceFile, and it'll be undefined and fail later.

This feels like it should be considered a bug in Node.  IMO if it's going to error on named imports because it doesn't understand how `module.exports` is constructed, then it should also error on a wildcard import instead of just silently returning an empty object.  That's a pretty big footgun; I don't imagine TS users are the only victims of it.

jakebailey commented:
> > But in native ESM, this is the empty object! You'll write ts.createSourceFile, and it'll be undefined and fail later.
> 
> This feels like it should be considered a bug in Node. IMO if it's going to error on named imports because it doesn't understand how `module.exports` is constructed, then it should also error on a wildcard import instead of just silently returning an empty object. That's a pretty big footgun; I don't imagine TS users are the only victims of it.

I think I over simplified; `.default` works, so it's not the empty object.

fatcerberus commented:
Even so, that’s still a pretty big footgun.  Anyone successfully doing a wildcard import from a CJS module probably expects to be able to do more than just `foo.default.bar()`, else they would just default-import in the first place.

jakebailey commented:
One problem with this approach is that live bindings don't actually work if you use named or namespace imports from native ESM. I'm not totally sure why, as my understanding was that they worked in esbuild thanks to all of the getters and such.

I can reproduce exactly the same behavior using esbuild's CJS output, so maybe this isn't a deal breaker and I'm misunderstanding things (live bindings work for the default export). Node also says that they don't update: https://nodejs.org/api/esm.html#:~:text=Live%20binding%20updates%20or%20new%20exports%20added%20to%20module.exports%20are%20not%20detected%20for%20these%20named%20exports.

It's not like we have any actual bindings that update after the fact, besides say, sys, which is our only mutable export for "reasons".

The default export is definitely a better thing to be doing at the user level, though, but I'm not sure how to statically force users of the TS API to do so (and people _definitely_ use a star import more often than a default).

jakebailey commented:
During the design meeting, @andrewbranch pointed to #54018, which funnily contains the same explanation I independently described (which I am going to place on the Somerton Scale as "subconscious appropriation" 😄)

jakebailey commented:
Evan pointed out that we could just flip our bundle from "global that's conditionally turned into CJS" to "CHS that's conditionally turned into a global", a la UMD. If we then switch to named exports, the annotations will be produced and the lexer will dig into the closure.

There are some gotchas but I'm going to draft this in the meantime.

jakebailey commented:
I used @evanw's suggestion in https://github.com/evanw/esbuild/issues/3607#issuecomment-1911523751 and pushed it here; the only gotcha is that pesky `__esModule` property, which I'm currently hacking away by replacing esbuild's helper that adds it (rather than copying the export object a second time).

The new bundle is now actually 5% faster to start up and 5% smaller (8.8MB to 8.3MB):

|  | Before | After | Diff | Diff (percent) |
| - | - | - | - | - |
| Packed | 5.55 MiB | 5.54 MiB | -16.86 KiB | -0.30% |
| Unpacked | 30.85 MiB | 30.35 MiB | -512.46 KiB | -1.62% |

|  | Before | After | Diff | Diff (percent) |
| - | - | - | - | - |
| `lib/typescript.js` | 8.71 MiB | 8.21 MiB | -512.46 KiB | -5.75% |

---

@typescript-bot perf test

jakebailey commented:
I'm undrafting this; the only concern is the hack for CJS, which I've mitigated by asserting that the replacement pattern must match exactly once.

That and the timing, given we just pushed out the beta and this does change the bundle quite a bit.

jakebailey commented:
@typescript-bot pack this

jakebailey commented:
Oof, this breaks monaco.

jakebailey commented:
@typescript-bot pack this

jakebailey commented:
Alright, it's fixed now. Arguably monaco is doing the wrong thing in their build by using `var ts` instead of the fake `module.exports` I added (my fault for not doing that when I fixed their codebase a while ago), but it's not hard to keep working in case someone else does this same thing.

jakebailey commented on src/compiler/core.ts:
```diff
@@ -2749,6 +2749,5 @@ export function isNodeLikeSystem(): boolean {
     return typeof process !== "undefined"
         && !!process.nextTick
-        && !(process as any).browser
-        && typeof module === "object";
```
In the new version, `module` is a parameter and is always provided. This check would no longer do anything. I'm not sure that it really matters that we don't have this.

jakebailey replied:
Actually, I'm going to switch this back to the same definition it used to be, e.g. checking for `require`. The comment I left above about `require` stopped being correct at some point as I refined our output bundle format to rewrite `require` temporarily.

jakebailey replied:
(I don't believe this will end up mattering in practice, but that's why we merge these sorts of things early in the cycle...)
